### PR TITLE
remember last N servers in setting

### LINKED ISSUE #148: Remember all successful tiled server connections
We could keep a list of all the successful tiled server connections in the user's settings file.  That list would populate the QComboBox.  While it is expected that most use would only use a single server, this is another _It would be nice if_ feature.

_Originally posted by @prjemian in https://github.com/BCDA-APS/gemviz/issues/143#issuecomment-1694398202_

### PR DESCRIPTION
**Pull request by rodolakis**
close #148

### CODE REVIEW

File: gemviz/mainwindow.py [hunk 1]
```diff
@@ -147,25 +143,24 @@ def clearContent(self, clear_cat=True):
...
-        if not uri_list:
-            previous_uri = settings.getKey(TILED_SERVER_SETTINGS_KEY)
-            candidate_uris = ["", previous_uri, TESTING_URL, LOCALHOST_URL, "Other..."]
+    def setServerList(self, selected_uri=None):
+        """Set the list of server URIs"""
+        recent_uris_str = settings.getKey(TILED_SERVER_SETTINGS_KEY)
+        recent_uris_list = recent_uris_str.split(",") if recent_uris_str else []
+        # fmt: off
+        if selected_uri and self.isValidServerUri(selected_uri):
+            final_uri_list = [selected_uri] + [uri for uri in recent_uris_list[: MAX_RECENT_URI - 1] if uri != selected_uri]
```
Comment (prjemian) [inline gemviz/mainwindow.py:152]: Easier to understand this code if the list comprehension is reformatted:

```py
final_uri_list = [selected_uri] + [
    uri
    for uri in recent_uris_list[: MAX_RECENT_URI - 1]
    if uri != selected_uri
]
```

File: gemviz/mainwindow.py [hunk 2]
```diff
@@ -147,25 +143,24 @@ def clearContent(self, clear_cat=True):
...
 
-    def setServerList(self, uri_list=None):
-        """Set the list of server URIs and remove duplicate"""
-        unique_uris = set()
-        new_server_list = []
-
-        if not uri_list:
-            previous_uri = settings.getKey(TILED_SERVER_SETTINGS_KEY)
-            candidate_uris = ["", previous_uri, TESTING_URL, LOCALHOST_URL, "Other..."]
+    def setServerList(self, selected_uri=None):
```
Comment (prjemian) [inline gemviz/mainwindow.py:146]: Now, this code does not take a list as argument and it does not set the list.  It adds the URI to the list.  The method name could be more representative of what it does: `addServerUri()`.  Update the docstring, too.

File: gemviz/mainwindow.py [hunk 3]
```diff
@@ -147,25 +143,24 @@ def clearContent(self, clear_cat=True):
...
+        """Set the list of server URIs"""
```
Comment (prjemian) [inline gemviz/mainwindow.py:147]: `"""Rebuild the list of server URIs."""`

**Comment by rodolakis** [comment]
Done.

**Review by prjemian — APPROVED** [review]
LGTM

**Comment by prjemian** [comment]
I resolved the merge commits.

**Comment by prjemian** [comment]
I believe this PR will resolve #148.  Added that to the PR summary above.

**Pull request merged**